src/storage/prisma.ts

```diff
@@ -289,32 +274,53 @@ export class PrismaStorage implements IDriveStorage {
             header: DocumentHeader;
             newState?: State<any, any> | undefined;
         }>
-    ) {
+    ): Promise<void> {
         let result: {
             operations: Operation[];
             header: DocumentHeader;
             newState?: State<any, any> | undefined;
         } | null = null;
 
-        await this.db.$transaction(
-            async tx => {
-                const document = await this.getDocument(drive, id, tx);
-                if (!document) {
-                    throw new Error(`Document with id ${id} not found`);
+        try {
+            await this.db.$transaction(
+                async tx => {
+                    const document = await this.getDocument(drive, id, tx);
+                    if (!document) {
+                        throw new Error(`Document with id ${id} not found`);
+                    }
+                    result = await callback(document);
+
+                    const { operations, header, newState } = result;
+                    return this._addDocumentOperations(
+                        tx,
+                        drive,
+                        id,
+                        operations,
+                        header
+                    );
+                },
+                {
+                    isolationLevel: 'ReadCommitted',
+                    maxWait: 10000,
+                    timeout: 20000
                 }
-                result = await callback(document);
-
-                const { operations, header, newState } = result;
-                return this._addDocumentOperations(
-                    tx,
+            );
+        } catch (e) {
+            // if optimistic update fails, retry with new state
+            if (
+                e instanceof PrismaClientKnownRequestError &&
+                e.code === 'P2025'
```

Comment:
This error is thrown by Prisma if the transaction failed because a Record was not found. This happens if the `version` condition fails.